File: src/main/kotlin/io/github/parzivalExe/guiApi/components/ExecuteCommandComponent.java
package io.github.parzivalExe.guiApi.components;

import io.github.parzivalExe.guiApi.Gui;
import io.github.parzivalExe.guiApi.antlr.interfaces.XMLContent;
import io.github.parzivalExe.guiApi.objects.Command;
import org.bukkit.Material;
import org.bukkit.entity.HumanEntity;
import org.bukkit.entity.Player;
import org.bukkit.event.inventory.ClickType;
import org.bukkit.event.inventory.InventoryAction;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collection;

@SuppressWarnings("unused")
public class ExecuteCommandComponent extends Component{

    @XMLContent(necessary = true)
    private ArrayList<Command> commands = new ArrayList<>();

    public ExecuteCommandComponent(ComponentMeta meta, ArrayList<Command> commands) {
        super(meta);
        this.commands = commands;
    }
    public ExecuteCommandComponent(ComponentMeta meta, Command command) {
        super(meta);
        if(command != null) {
            this.commands.add(command);
        }
    }
    public ExecuteCommandComponent(ComponentMeta meta) {
        this(meta, (Command) null);
    }
    public ExecuteCommandComponent() {
        this(new ComponentMeta("", new ItemStack(Material.WOOL)));
    }

    @Override
    public void componentClicked(@NotNull HumanEntity whoClicked, @NotNull Gui gui, @NotNull InventoryAction action, int slot, @NotNull ClickType clickType) {
        if(whoClicked instanceof Player) {

            //noinspection ConstantConditions
            if(commands != null || commands.size() > 0) {
                for(Command command : commands) {
                    ((Player) whoClicked).performCommand(command.buildCommand());
                }
            }else{
                whoClicked.sendMessage("This component has no Commands implemented right now");
            }

        }else{
            whoClicked.sendMessage("Normally, only players can click Components");
        }
        gui.closeGui();
    }

    public void addCommand(Command command) {
        this.commands.add(command);
    }
    public void addCommands(Collection<Command> commands) {
        this.commands.addAll(commands);
    }

    public void setCommands(ArrayList<Command> commands) {
        this.commands = commands;
    }

    public void clearCommands() {
        commands.clear();
    }

    public ArrayList<Command> getCommands() {
        return commands;
    }
}

File: src/main/java/io/github/parzivalExe/guiApi/components/ExecuteCommandComponent.kt
package io.github.parzivalExe.guiApi.components

import io.github.parzivalExe.guiApi.Gui
import io.github.parzivalExe.guiApi.GuiApiInitializer
import io.github.parzivalExe.guiApi.antlr.interfaces.XMLContent
import io.github.parzivalExe.guiApi.objects.Command
import org.bukkit.Material
import org.bukkit.entity.HumanEntity
import org.bukkit.entity.Player
import org.bukkit.event.inventory.ClickType
import org.bukkit.event.inventory.InventoryAction
import org.bukkit.inventory.ItemStack

@Suppress("unused", "MemberVisibilityCanBePrivate")
class ExecuteCommandComponent(meta: ComponentMeta, @XMLContent(necessary = true) val commands: ArrayList<Command>)
    : Component(meta) {


    constructor(meta: ComponentMeta, command: Command) : this(meta, arrayListOf(command))

    constructor(meta: ComponentMeta) : this(meta, arrayListOf())

    @Deprecated("DON'T USE: This Constructor is only used for XML and shouldn't be used in Code itself", ReplaceWith("new ExecuteCommandComponent(ComponentMeta)"))
    constructor() : this(ComponentMeta("", ItemStack(Material.WHITE_WOOL)))


    override fun componentClicked(whoClicked: HumanEntity, gui: Gui, action: InventoryAction, slot: Int, clickType: ClickType) {
        if(whoClicked !is Player) {
            whoClicked.sendMessage("${GuiApiInitializer.PREFIX} Normally, only players can click Components")
            return
        }
        if(commands.isEmpty()) {
            whoClicked.sendMessage("${GuiApiInitializer.C_PREFIX} This component has no Commands implemented right now")
            return
        }

        for(command in commands) {
            whoClicked.performCommand(command.buildCommand())
        }
        gui.closeGui()
    }


    fun addCommand(command: Command) {
        commands += command
    }
    fun addCommands(commands: Collection<Command>) {
        this.commands += commands
    }

    fun setCommands(commands: ArrayList<Command>) {
        clearCommands()
        this.commands += commands
    }

    fun clearCommands() = commands.clear()

}